---
title: "INTERVAL Misexpression Fig.2 and associated supplementary Figs"
output: 
---

```{r}
library(ggplot2)
library(dplyr) 
library(RColorBrewer)
```

```{r}
### variables 
ticks_size <- 0.75
border_size <- 1.5
z_cutoff_to_show <- 10
z_cutoffs_to_show <- c(2, 10, 20, 30, 40)
z_cutoffs_names <- c("> 2", "> 10", "> 20", "> 30", "> 40")

### color scheme 
snp_indel_sv_vrnt_colors <- c("snp"="#DC853E", #DDCC77", 
                           "indel"="#44AA99", #'#44AA99', 
                           "all_sv"="#332288")

sv_vrnt_colors <- c("INV" = "#0081C1", #'#CC6677', #5CC9FF
                    'DEL' = '#E04B4B', 
                    'DUP' ='#00C750',
                    'MEI' = '#AA4499' 
                    )

snv_indel_sv_labeller <- c("snp"="SNV",  
                           "indel"="Indel", 
                           "all_sv"="SV")

sv_class_labeller <- c("DEL"="Deletions", 
                       "DUP"="Duplications", 
                       "INV"="Inversions", 
                       "MEI"="Mobile element insertions"
                       )
```

```{r}
# output directory
wkdir <- "/lustre/scratch126/humgen/projects/interval_rna/interval_rna_seq/thomasVDS/misexpression_v3"
# Fig. 2
fig2_out_dir <- file.path(wkdir, "figs/fig_2")
dir.create(file.path(fig2_out_dir))
# Supplementary Fig. 5
supp_fig5_out_dir <- file.path(wkdir, "figs/supp_fig_5")
dir.create(file.path(supp_fig5_out_dir))
# Supplementary Fig. 6
supp_fig6_out_dir <- file.path(wkdir, "figs/supp_fig_6")
dir.create(file.path(supp_fig6_out_dir))
# Supplementary Fig. 7
supp_fig7_out_dir <- file.path(wkdir, "figs/supp_fig_7")
dir.create(file.path(supp_fig7_out_dir))
# Supplementary Fig. 8
supp_fig8_out_dir <- file.path(wkdir, "figs/supp_fig_8")
dir.create(file.path(supp_fig8_out_dir))
```

### SNV, indel, SV MAF and z-score cutoffs 
```{r}
theme.snv_indel_sv_maf_z <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_rect(color = "black", fill = NA, size = border_size),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_line(size=ticks_size),
          axis.ticks.length = unit(.15, "cm"),
          axis.title.x = element_text(size=base_size+2, family=base_family, face="plain", margin = unit(c(2, 0, 0, 0), "mm"), color="black"),
          axis.title.y = element_text(size=base_size+2, family=base_family, face="plain", angle=90, margin = unit(c(0, 2, 0, 0), "mm"), color="black"),
          axis.text.x = element_text(size=base_size+2, family=base_family, face="plain", margin = unit(c(1, 0, 0, 0), "mm"), color="black"), 
          axis.text.y = element_text(size=base_size+2, family=base_family, face="plain", margin = unit(c(0, 1, 0, 0), "mm"), color="black"), 
          legend.key = element_blank(),
          legend.title = element_blank(),
          legend.text = element_text(size=base_size+2, family = base_family, face="plain", color="black"),
          legend.justification = c(0.98, 0.98),
          legend.position = c(0.98, 0.98),
          complete = TRUE,
          strip.placement = "outside",
          strip.background = element_blank(),
          strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0.5, vjust=1),
          plot.title = element_text(hjust=0.5, size=base_size+2, family=base_family, face="plain", color="black"))
  return(p)
}

snv_indel_sv_results_path <- file.path(wkdir, "4_vrnt_enrich/enrich_results_mul_test/grouped_enrich/snp_indel_sv_gene_window_results.tsv")
snv_indel_sv_results_df <- read.csv(snv_indel_sv_results_path, sep="\t")

# factor for MAF range
snv_indel_sv_results_df$maf_range <- as.character(snv_indel_sv_results_df$maf_range)
snv_indel_sv_results_df$maf_range <- factor(snv_indel_sv_results_df$maf_range, levels=unique(snv_indel_sv_results_df$maf_range))
# factor for z-score cutoffs 
snv_indel_sv_z_cutoff_results_df <- subset(snv_indel_sv_results_df, snv_indel_sv_results_df$z_cutoff %in% z_cutoffs_to_show)
snv_indel_sv_z_cutoff_results_df$z_cutoff_name <- as.character(snv_indel_sv_z_cutoff_results_df$z_cutoff_name)
snv_indel_sv_z_cutoff_results_df$z_cutoff_name <- factor(snv_indel_sv_z_cutoff_results_df$z_cutoff_name, levels=z_cutoffs_names)

facet_labels <- c("0-1" = "Rare\n 0-1%",
                  "1-5" = "Low-frequency\n 1-5%",
                  "5-10" = "Common\n 5-10%",
                  "10-50"= "Common\n 10-50%"
                  )

snv_indel_sv_z_nolog_path <- file.path(fig2_out_dir, "snv_indel_sv_enrich_gene_body.pdf")
pdf(snv_indel_sv_z_nolog_path, width = 9,height = 3.5, useDingbats=FALSE)

snv_indel_sv_enrich_gene_body <- ggplot(snv_indel_sv_z_cutoff_results_df, aes(x=z_cutoff_name, y=risk_ratio, color=vrnt_type)) + 
  geom_hline(yintercept=1.0, linetype="solid", size=ticks_size) + 
  geom_point(position = position_dodge(width = 0.6), size=3) + 
  geom_errorbar(aes(ymin=risk_ratio_lower, ymax=risk_ratio_upper),cex=1, width=0, position = position_dodge(width=0.6)) + 
  scale_x_discrete(name = "Misexpression z-score threshold", labels=c("2", "10", "20", "30", "40", "50")) +
  scale_y_continuous(name = "Enrichment", breaks=c(1, 3, 5, 7, 9), limits=c(0, 8.5), expand=c(0,0)) +
  scale_color_manual(name = '', 
                    labels=snv_indel_sv_labeller,
                    values= snp_indel_sv_vrnt_colors, 
                    ) +
  facet_grid(.~maf_range, scales = "free", space = "free", switch='y', labeller = as_labeller(facet_labels)) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio > 1, "*", ""), y=risk_ratio_upper), hjust = 0.5, vjust = -0.25, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio < 1, "*", ""), y=risk_ratio_lower), hjust = 0.5, vjust = 1.75, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL)))

theme.snv_indel_sv_maf_z(snv_indel_sv_enrich_gene_body)
dev.off()
theme.snv_indel_sv_maf_z(snv_indel_sv_enrich_gene_body)
snv_indel_sv_z_path_nolog_png <- file.path(fig2_out_dir, "snv_indel_sv_enrich_gene_body.png")
ggsave(snv_indel_sv_z_path_nolog_png, width=9, height=3.5, dpi=360)
```

### Percentage misexpression carriers at different z-score cuttoffs
```{r}
### Inactive gene set QC summary 
perc_misexp_carrier_path <- file.path(wkdir, "4_vrnt_enrich/combine_count_carriers/perc_misexp_carrier_zscores.tsv")
perc_misexp_carrier_df <- read.csv(perc_misexp_carrier_path, sep="\t")

perc_misexp_carrier_df$z.score <- factor(perc_misexp_carrier_df$z.score, levels=z_cutoffs_names)

# save figure 
width=6
height=5
perc_misexp_carrier_zscore_pdf = file.path(supp_fig5_out_dir, "perc_misexp_carrier_zscore.pdf")
pdf(file=perc_misexp_carrier_zscore_pdf, width = width, height = height, useDingbats=FALSE)

theme.perc_misexp_carrier <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(color="black", size=0.75),
          axis.ticks.y = element_line(size=0.75),
          axis.ticks.x = element_blank(),
          axis.ticks.length = unit(.15, "cm"),
          axis.title = element_text(size=base_size+2, family=base_family, face="plain", color="black"),
          axis.text = element_text(size=base_size+2, family=base_family, face="plain", color="black"),
          strip.background = element_blank(),
          legend.key = element_blank(),
          legend.title = element_text(size=base_size, family = base_family, face="bold"),
          legend.text = element_text(size=base_size - 1, family = base_family, face="plain"),
          complete = TRUE,
          plot.title = element_text(hjust=0.5))
  return(p)
}

perc_misexp_carrier_fig <- ggplot(perc_misexp_carrier_df, aes(y=perc_carrier, x=z.score)) +
   geom_bar(stat = "identity", fill = "#5B84B1FF", alpha = 0.5, color="black", size=ticks_size) + 
   scale_y_continuous(name = "% misexpression events with SV within 200 kb", limits=c(0,6), expand=c(0,0)) + 
   scale_x_discrete(name = "Misexpression z-score threshold", expand=c(0.1, 0.1)) + 
   geom_text(aes(label = round(misexp_carrier, 1)), vjust =-0.5, size = 4.5, fontface="plain", color="black")
theme.perc_misexp_carrier(perc_misexp_carrier_fig)
dev.off()
theme.perc_misexp_carrier(perc_misexp_carrier_fig)
perc_misexp_carrier_zscore_png <- file.path(supp_fig5_out_dir, "perc_misexp_carrier_zscore.png")
ggsave(perc_misexp_carrier_zscore_png, width=width, height=height, dpi=360)
```
### SNV, indel and SV variant enrichment across windows 
```{r}
theme.snp_indel_sv_window <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_rect(color = "black", fill = NA, size = border_size),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_blank(),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_line(size=ticks_size),
          axis.ticks.length = unit(.15, "cm"),
          axis.title.x = element_blank(), #element_text(size=base_size+2, family=base_family, face="plain", color="black"),
          axis.title.y = element_text(size=base_size+2, family=base_family, face="plain", angle=90, color="black", margin = unit(c(0, 2, 0, 0), "mm")),
          axis.text.x = element_blank(), #element_text(size=base_size+2, family=base_family, face="plain", angle=90, hjust=1, color="black"),
          axis.text.y = element_text(size=base_size+2, family=base_family, face="plain",color="black", margin = unit(c(0, 1, 0, 0), "mm")),
          legend.key = element_blank(),
          legend.title = element_text(size=base_size+2, family = base_family, face="plain",color="black"),
          legend.text = element_text(size=base_size+2, family = base_family, face="plain",color="black"),
          complete = TRUE,
          strip.placement = "outside",
          strip.background = element_blank(),
          strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0.5, angle=0, vjust = 1),
          plot.title = element_text(hjust=0.5))
  return(p)
}

snp_indel_sv_window_bins_path <- file.path(wkdir,"4_vrnt_enrich/enrich_results_mul_test/grouped_enrich/snp_indel_sv_maf_0_1perc_window_bin_results.tsv")
snp_indel_sv_window_bins_df <- read.csv(snp_indel_sv_window_bins_path, sep="\t")
# factor for z-score cutoffs 
snp_indel_sv_window_bins_z_subset_df <- subset(snp_indel_sv_window_bins_df, snp_indel_sv_window_bins_df$z_cutoff %in% z_cutoffs_to_show)
snp_indel_sv_window_bins_z_subset_df$z_cutoff_name <- as.character(snp_indel_sv_window_bins_z_subset_df$z_cutoff_name)
snp_indel_sv_window_bins_z_subset_df$z_cutoff_name <- factor(snp_indel_sv_window_bins_z_subset_df$z_cutoff_name, levels=z_cutoffs_names)
# factor for windows
level_order <- c("-800kb to -1Mb", "-600kb to -800kb", "-400kb to -600kb", "-200kb to -400kb", "TSS to -200kb", 
                 "gene body", "TTS to 200kb", "200kb to 400kb", "400kb to 600kb", "600kb to 800kb", "800kb to 1Mb")
snp_indel_sv_window_bins_z_subset_df$vrnt_type_f = factor(snp_indel_sv_window_bins_z_subset_df$vrnt_type, levels=c('snp','indel','all_sv'))
snp_indel_sv_window_bins_z_cutoff_df = subset(snp_indel_sv_window_bins_z_subset_df, z_cutoff == z_cutoff_to_show)


width = 4.5
height = 4
snv_indel_sv_z_window_path = file.path(fig2_out_dir, "snv_indel_sv_enrich_windows_stacked.pdf")
pdf(snv_indel_sv_z_window_path, width = width, height = height, useDingbats=FALSE)

snv_indel_sv_z_window <- ggplot(snp_indel_sv_window_bins_z_cutoff_df, aes(x=factor(window_name, level=level_order), y=risk_ratio, color=vrnt_type)) + 
  geom_hline(yintercept=1.0, linetype="solid", size=ticks_size) + 
  geom_point(position = position_dodge(width = 0.6), size=3) +
  ggtitle("") +
  geom_errorbar(aes(ymin=risk_ratio_lower, ymax=risk_ratio_upper),cex=1, width=0, position = position_dodge(width = 0.6)) + 
  scale_x_discrete(name = "Genomic window") + #, labels=x_axis_labels) +
  scale_y_continuous(name = "Enrichment", 
                   limits=c(-1.5, 9), 
                   breaks=c( 1, 3, 5, 7), 
                   expand=c(0,0)
                  ) + 
  scale_colour_manual(name = 'Variant\n  Type', 
                    values= snp_indel_sv_vrnt_colors, 
                    labels=snv_indel_sv_labeller,
                    guide="none"
                    ) +
  facet_wrap(~vrnt_type_f, ncol = 1, as.table=TRUE, scales = "free_y", labeller=as_labeller(snv_indel_sv_labeller)) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio > 1, "*", ""), y=risk_ratio_upper), hjust = 0.5, vjust = -0.25, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio < 1, "*", ""), y=risk_ratio_lower), hjust = 0.5, vjust = 1.75, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL)))

theme.snp_indel_sv_window(snv_indel_sv_z_window)
dev.off()
theme.snp_indel_sv_window(snv_indel_sv_z_window)
snv_indel_sv_z_window_path_png <- file.path(fig2_out_dir, "snv_indel_sv_enrich_windows_stacked.png")
ggsave(snv_indel_sv_z_window_path_png, width = width, height = height, dpi=360)
```

### SNV, indel and SV variant enrichment across windows and multiple z-score thresholds 
```{r}
### Windows supplementary figure 
theme.misexp_snv_indel_windows_z <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_rect(color = "black", fill = NA, size = 1),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_line(size=ticks_size),
          axis.ticks.length = unit(.15, "cm"),
          axis.title.x = element_text(size=base_size+2, family=base_family, face="plain", color="black"),
          axis.title.y = element_text(size=base_size+2, family=base_family, face="plain", angle=90, color="black", margin = unit(c(0, 2, 0, 0), "mm")),
          axis.text.x = element_text(size=base_size+2, family=base_family, face="plain", angle=90, hjust=1, color="black"),
          axis.text.y = element_text(size=base_size+2, family=base_family, face="plain",color="black", margin = unit(c(0, 1, 0, 0), "mm")),
          legend.key = element_blank(),
          legend.title = element_text(size=base_size+2, family = base_family, face="plain",color="black"),
          legend.text = element_text(size=base_size+2, family = base_family, face="plain",color="black"),
          complete = TRUE,
          strip.placement = "outside",
          strip.background = element_blank(),
          strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0.5, vjust=1),
          strip.text.y = element_text(size=base_size+2, family = base_family, face="plain", hjust=0.5, vjust=0.5, angle=0),
          plot.title = element_text(hjust=0.5))
  return(p)
}

snv_indel_sv_z_window_z_cutoffs_path = file.path(supp_fig6_out_dir, "snv_indel_sv_enrich_windows_z_cutoffs.pdf")
pdf(snv_indel_sv_z_window_z_cutoffs_path, width=9, height=10, useDingbats=FALSE)

snv_indel_sv_window_z_cutoffs <- ggplot(snp_indel_sv_window_bins_z_subset_df, aes(x=factor(window_name, level=level_order), y=risk_ratio, color=vrnt_type)) + 
  geom_hline(yintercept=1.0, linetype="solid", size=ticks_size) + 
geom_point(position = position_dodge(width = 0.6), size=3) +
ggtitle("") +
geom_errorbar(aes(ymin=risk_ratio_lower, ymax=risk_ratio_upper),cex=1, width=0, position = position_dodge(width = 0.6)) + 
scale_x_discrete(name = "Genomic window") + #, labels=x_axis_labels) +
scale_y_continuous(name = "Enrichment", 
                   expand=c(0.2,0.2)
                  ) + 
scale_colour_manual(name = 'Variant\n  Type', 
                    values= snp_indel_sv_vrnt_colors, 
                    labels=snv_indel_sv_labeller,
                    guide="none") +
  facet_grid(z_cutoff_name~., scales="free_y", space="fixed") + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio > 1, "*", ""), y=risk_ratio_upper), hjust = 0.5, vjust = -0.25, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio < 1, "*", ""), y=risk_ratio_lower), hjust = 0.5, vjust = 1.75, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL)))

theme.misexp_snv_indel_windows_z(snv_indel_sv_window_z_cutoffs)
dev.off()
theme.misexp_snv_indel_windows_z(snv_indel_sv_window_z_cutoffs)
snv_indel_sv_enrich_windows_z_cutoffs_png <- file.path(supp_fig6_out_dir, "snv_indel_sv_enrich_windows_z_cutoffs.png")
ggsave(snv_indel_sv_enrich_windows_z_cutoffs_png, width=9, height=10, dpi=360)
```

### Enrichment of rare DEL and DUP near to misexpressed genes  
```{r}
theme.sv_class_enrich <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_rect(color = "black", fill = NA, size = border_size),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_line(size=ticks_size),
          axis.ticks.length = unit(.15, "cm"),
          axis.title.x = element_text(size=base_size+2, family=base_family, face="plain", vjust=-1, margin = unit(c(2, 0, 0, 0), "mm"), color="black"),
          axis.title.y = element_text(size=base_size+2, family=base_family, face="plain", angle=90, margin = unit(c(0, 2, 0, 0), "mm"), color="black"),
          axis.text.y = element_text(size=base_size+2, family=base_family, face="plain", margin = unit(c(0, 1, 0, 0), "mm"), color="black"),
          axis.text.x = element_text(size=base_size+2, family=base_family, face="plain", hjust=0.5, margin = unit(c(1, 0, 0, 0), "mm"), color="black"),
          
          legend.key = element_blank(),
          legend.title = element_text(size=base_size+2, family = base_family, face="bold"),
          legend.text = element_text(size=base_size+2, family = base_family, face="plain"),
          complete = TRUE,
          strip.placement = "outside",
          strip.background = element_blank(),
          strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0.5, angle=0, vjust = 1),
          plot.title = element_text(hjust=0.5, size=base_size+2, family=base_family, face="bold"),
          panel.spacing = unit(0.2, "lines"))
  return(p)
}

sv_types_enrich_results_all_path = file.path(wkdir, "4_vrnt_enrich/enrich_results_mul_test/grouped_enrich/sv_types_maf_0_1perc_200kb_all_z_results.tsv")
sv_types_enrich_results_all_df <- read.csv(sv_types_enrich_results_all_path, sep="\t")

# factor z-score cutoffs 
sv_types_z_cutoff_subset_results_df <- subset(sv_types_enrich_results_all_df, sv_types_enrich_results_all_df$z_cutoff %in% z_cutoffs_to_show)
sv_types_z_cutoff_subset_results_df$z_cutoff_name <- as.character(sv_types_z_cutoff_subset_results_df$z_cutoff_name)
sv_types_z_cutoff_subset_results_df$z_cutoff_name <- factor(sv_types_z_cutoff_subset_results_df$z_cutoff_name, levels=z_cutoffs_names)

# restrict to DEL and DUPs 
vrnt_types_to_include = c("DEL", "DUP")
del_dup_z_cutoff_subset_results_df <- subset(sv_types_z_cutoff_subset_results_df, sv_types_z_cutoff_subset_results_df$vrnt_type %in% vrnt_types_to_include)

width = 4.5
height = 4
rare_sv_type_enrich_path = file.path(fig2_out_dir, "rare_del_dup_enrich_gene_body.pdf")
pdf(rare_sv_type_enrich_path, width = width,height = height, useDingbats=FALSE)

sv_class_enrich <- ggplot(subset(del_dup_z_cutoff_subset_results_df, del_dup_z_cutoff_subset_results_df$pval < 0.05), aes(y=risk_ratio, x=z_cutoff_name, color=vrnt_type)) + 
    geom_hline(yintercept=1.0, linetype="solid", size=ticks_size) + 
  geom_point(size=3) + 
  geom_errorbar(aes(ymin=risk_ratio_lower, ymax=risk_ratio_upper), cex=1, width=0) + 
  scale_y_continuous(name = "Enrichment", expand=c(0.2, 0.2)) +
  scale_x_discrete(name = "Misexpression z-score threshold", labels=c("2", "10", "20", "30", "40")) +
  facet_wrap(vrnt_type~., ncol = 1, as.table=TRUE, labeller=as_labeller(sv_class_labeller), scales="free_y") + 
  scale_colour_manual(name = "", values = sv_vrnt_colors, guide="none") + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio > 1, "*", ""), y=risk_ratio_upper), hjust = 0.5, vjust = -0.25, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio < 1, "*", ""), y=risk_ratio_lower), hjust = 0.5, vjust = 1.75, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL)))

theme.sv_class_enrich(sv_class_enrich)
dev.off()
theme.sv_class_enrich(sv_class_enrich)
sv_class_enrich_path_png <- file.path(fig2_out_dir, "rare_del_dup_enrich_gene_body.png")
ggsave(sv_class_enrich_path_png, width = width, height = height, dpi=360)
```

### Enrichment of rare INV and MEI near to misexpressed genes 
```{r}
vrnt_types_to_include = c("INV", "MEI")
inv_mei_z_cutoff_subset_results_df <- subset(sv_types_z_cutoff_subset_results_df, sv_types_z_cutoff_subset_results_df$vrnt_type %in% vrnt_types_to_include)

width = 8 
height = 4
rare_inv_mei_type_enrich_path = file.path(supp_fig7_out_dir, "rare_inv_mei_enrich_gene_body.pdf")
pdf(rare_inv_mei_type_enrich_path, width = width, height = height, useDingbats=FALSE)

inv_mei_enrich <- ggplot(inv_mei_z_cutoff_subset_results_df, aes(y=risk_ratio, x=z_cutoff_name, color=vrnt_type)) + 
    geom_hline(yintercept=1.0, linetype="solid", size=ticks_size) + 
  geom_point(size=3) + 
  geom_errorbar(aes(ymin=risk_ratio_lower, ymax=risk_ratio_upper), cex=1, width=0) + 
  scale_y_continuous(name = "Enrichment", expand=c(0.1, 0.1)) +
  scale_x_discrete(name = "Misexpression z-score threshold", labels=c("2", "10", "20", "30", "40")) +
  facet_wrap(vrnt_type~., ncol = 2, as.table=TRUE, labeller=as_labeller(sv_class_labeller)) + 
  scale_colour_manual(name = "", values = sv_vrnt_colors, guide="none") + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio > 1, "*", ""), y=risk_ratio_upper), hjust = 0.5, vjust = -0.25, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio < 1, "*", ""), y=risk_ratio_lower), hjust = 0.5, vjust = 1.75, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL)))

theme.sv_class_enrich(inv_mei_enrich)
dev.off()
theme.sv_class_enrich(inv_mei_enrich)
rare_inv_mei_type_enrich_png <- file.path(supp_fig7_out_dir, "rare_inv_mei_enrich_gene_body.png")
ggsave(rare_inv_mei_type_enrich_png, width = width, height = height, dpi=600)
```

### SV type window enrichment: DEL, DUP, INV
```{r}
theme.sv_class_window_enrich <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_rect(color = "black", fill = NA, size = border_size),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_line(size=ticks_size),
          axis.ticks.length = unit(.15, "cm"),
          axis.title.x = element_text(size=base_size+2, family=base_family, face="plain", vjust=-1, margin = unit(c(2, 0, 0, 0), "mm"), color="black"),
          axis.title.y = element_text(size=base_size+2, family=base_family, face="plain", angle=90, margin = unit(c(0, 2, 0, 0), "mm"), color="black"),
          axis.text.y = element_text(size=base_size+2, family=base_family, face="plain", margin = unit(c(0, 1, 0, 0), "mm"), color="black"),
          axis.text.x = element_text(size=base_size+2, family=base_family, face="plain", hjust=1, margin = unit(c(1, 0, 0, 0), "mm"), color="black", angle=90),
          legend.key = element_blank(),
          legend.title = element_text(size=base_size+2, family = base_family, face="plain"),
          legend.text = element_text(size=base_size+2, family = base_family, face="plain"),
          complete = TRUE,
          strip.placement = "outside",
          strip.background = element_blank(),
          #strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0, angle=0, vjust=1),
          plot.title = element_text(size=base_size, family=base_family, face="bold", hjust=0, angle=0, vjust = 1),
          strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0.5, angle=0, vjust = 1),
          panel.spacing = unit(0.2, "lines"))
  return(p)
}

sv_types_window_enrich_results_path <- file.path(wkdir, "4_vrnt_enrich/enrich_results_mul_test/grouped_enrich/sv_types_maf_0_1perc_windows_all_z_results.tsv")
sv_types_window_enrich_results_df <- read.csv(sv_types_window_enrich_results_path, sep="\t") 
window_labels_ordered = c("-800kb to -1Mb", "-600kb to -800kb", "-400kb to -600kb", "-200kb to -400kb", "TSS to -200kb", "gene body", 
                          "TTS to 200kb", "200kb to 400kb", "400kb to 600kb", "600kb to 800kb", "800kb to 1Mb")    

sv_types_window_enrich_results_df$window_name <- as.character(sv_types_window_enrich_results_df$window_name)
sv_types_window_enrich_results_df$window_name <- factor(x = sv_types_window_enrich_results_df$window_name, levels = window_labels_ordered)

width = 4.5
height = 5.5
rare_sv_type_enrich_windows_path = file.path(fig2_out_dir, "rare_sv_types_enrich_windows.pdf")
pdf(rare_sv_type_enrich_windows_path, width = width, height = height, useDingbats=FALSE)

sv_class_window_enrich_fig <- ggplot(subset(sv_types_window_enrich_results_df, z_cutoff == z_cutoff_to_show & vrnt_type != "MEI"), aes(y=risk_ratio, x=window_name, color=vrnt_type)) + 
  geom_hline(yintercept=1.0, linetype="solid", size=ticks_size) + 
  geom_point(size=3) + 
  ggtitle("") +
  geom_errorbar(aes(ymin=risk_ratio_lower, ymax=risk_ratio_upper),cex=1, width=0) + 
  scale_y_continuous(name = "Enrichment", expand=c(0.42, 0.42)) + #, labels = c(0.1, 1, 10, 100), breaks= c(0.1, 1, 10, 100)) +
  scale_x_discrete(name = "Genomic window", labels=window_labels_ordered) +
  facet_wrap(~vrnt_type, ncol = 1, as.table=TRUE,scales = "free_y", labeller=as_labeller(sv_class_labeller)) + 
  scale_colour_manual(name = "", 
                      values =sv_vrnt_colors, 
                      guide="none") + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio > 1, "*", ""), y=risk_ratio_upper), hjust = 0.5, vjust = -0.25, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio < 1, "*", ""), y=risk_ratio_lower), hjust = 0.5, vjust = 1.75, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL)))

theme.sv_class_window_enrich(sv_class_window_enrich_fig)
dev.off()
theme.sv_class_window_enrich(sv_class_window_enrich_fig)
rare_sv_type_enrich_windows_path_png <- file.path(fig2_out_dir, "rare_sv_types_enrich_windows.png")
ggsave(rare_sv_type_enrich_windows_path_png, width = width, height = height, dpi=600)
```

### INV, DEL and DUP windows enrichment across z-scores 
```{r}
theme.sv_class_window_enrich_supp <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_rect(color = "black", fill = NA, size = border_size),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_line(size=ticks_size),
          axis.ticks.length = unit(.15, "cm"),
          axis.title.x = element_text(size=base_size+2, family=base_family, face="plain", vjust=-1, margin = unit(c(2, 0, 0, 0), "mm"), color="black"),
          axis.title.y = element_text(size=base_size+2, family=base_family, face="plain", angle=90, margin = unit(c(0, 2, 0, 0), "mm"), color="black"),
          axis.text.y = element_text(size=base_size+2, family=base_family, face="plain", margin = unit(c(0, 1, 0, 0), "mm"), color="black"),
          axis.text.x = element_text(size=base_size+2, family=base_family, face="plain", hjust=1, margin = unit(c(1, 0, 0, 0), "mm"), color="black", angle=90),
          legend.key = element_blank(),
          legend.title = element_text(size=base_size+2, family = base_family, face="plain"),
          legend.text = element_text(size=base_size+2, family = base_family, face="plain"),
          complete = TRUE,
          strip.placement = "outside",
          strip.background = element_blank(),
          #strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0, angle=0, vjust=1),
          plot.title = element_text(size=base_size, family=base_family, face="bold", hjust=0, angle=0, vjust = 1),
          strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0.5, angle=0, vjust = 1),
          panel.spacing = unit(0.2, "lines"))
  return(p)
}

sv_types_window_enrich_results_z_cutoffs_df <- subset(sv_types_window_enrich_results_df, sv_types_window_enrich_results_df$z_cutoff %in% c(2, 10, 20, 30, 40) & sv_types_window_enrich_results_df$vrnt_type != "MEI")
sv_types_window_enrich_results_z_cutoffs_df$z_cutoff_name <- factor(sv_types_window_enrich_results_z_cutoffs_df$z_cutoff_name, levels=c("> 2", "> 10", "> 20", "> 30", "> 40"))

rare_sv_type_enrich_z_cutoffs_path = file.path(supp_fig7_out_dir, "rare_sv_types_enrich_windows_z_cutoffs.pdf")
pdf(rare_sv_type_enrich_z_cutoffs_path, width = 12,height = 10, useDingbats=FALSE)

sv_class_window_labeller <- c("DEL"="Deletions", 
                              "DUP"="Duplications", 
                              "INV"="Inversions", 
                              "> 2" = " > 2",
                              "> 10" = "> 10",
                              "> 20" = "> 20",
                              "> 30" = "> 30", 
                              "> 40" = "> 40"
                       )
risk_ratio_max = max(sv_types_window_enrich_results_z_cutoffs_df$risk_ratio_upper) + 1000

sv_class_window_enrich_fig <- ggplot(sv_types_window_enrich_results_z_cutoffs_df, aes(y=risk_ratio, x=window_name, color=vrnt_type)) + 
  geom_hline(yintercept=1.0, linetype="solid", size=ticks_size) + 
  geom_point(size=3) + 
  ggtitle("") +
  geom_errorbar(aes(ymin=risk_ratio_lower, ymax=risk_ratio_upper),cex=1, width=0) + 
  scale_y_continuous(name = "Enrichment", expand=c(0.2, 0.2)) +
  scale_x_discrete(name = "Genomic window", labels=window_labels_ordered) +
  facet_grid(vrnt_type~z_cutoff_name,  scales = "free", as.table=TRUE, labeller=as_labeller(sv_class_window_labeller)) + 
  scale_colour_manual(name = "", 
                      values =sv_vrnt_colors, guide="none",) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio > 1, "*", ""), y=risk_ratio_upper), hjust = 0.5, vjust = -0.25, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio < 1, "*", ""), y=risk_ratio_lower), hjust = 0.5, vjust = 1.75, position = position_dodge(width = 0.6), size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL))) 

theme.sv_class_window_enrich_supp(sv_class_window_enrich_fig)
dev.off()
theme.sv_class_window_enrich_supp(sv_class_window_enrich_fig)
rare_sv_type_enrich_z_cutoffs_path_png <- file.path(supp_fig7_out_dir, "rare_sv_types_enrich_windows_z_cutoffs.png")
ggsave(rare_sv_type_enrich_z_cutoffs_path_png, width=12, height=10, dpi=360)
```

### All SV consequence that pass Bonferroni cutoff 
```{r}
### SV consequence that pass Bonferroni cutoff 
theme.sv_class_enrich <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_rect(color = "black", fill = NA, size = border_size),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_line(size=ticks_size),
          axis.ticks.length = unit(.15, "cm"),
          axis.title.x = element_text(size=base_size+2, family=base_family, face="plain", vjust=-1, margin = unit(c(2, 0, 0, 0), "mm"), color="black"),
          axis.title.y = element_text(size=base_size+2, family=base_family, face="plain", angle=90, margin = unit(c(0, 2, 0, 0), "mm"), color="black"),
          axis.text.y = element_text(size=base_size+2, family=base_family, face="plain", hjust=1, margin = unit(c(0, 1, 0, 0), "mm"), color="black"),
          axis.text.x = element_text(size=base_size+2, family=base_family, face="plain", hjust=0.5, margin = unit(c(1, 0, 0, 0), "mm"), color="black"),
          legend.key = element_blank(),
          legend.title = element_text(size=base_size+2, family = base_family, face="bold"),
          legend.text = element_text(size=base_size+2, family = base_family, face="plain"),
          legend.position = "none",
          complete = TRUE,
          strip.placement = "outside",
          strip.background = element_blank(),
          strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0.5, angle=0, vjust = 1),
          plot.title = element_text(hjust=0.5, size=base_size+2, family=base_family, face="bold"),
          panel.spacing = unit(0.2, "lines"))
  return(p)
}
# z-cutoff 
sv_msc_enrich_results_path <- file.path(wkdir,'4_vrnt_enrich/enrich_results_mul_test/grouped_enrich/sv_consq_pass_bonf_z10.tsv')
sv_msc_enrich_results_df = read.csv(sv_msc_enrich_results_path, sep="\t", stringsAsFactors = FALSE)
# order consequences 
sv_consequence_order <- rev(unique(c('Upstream', 'Non-coding transcript','Coding','Transcript amplification','No predicted effect')))
sv_msc_enrich_results_df$consq_name_f <- as.character(sv_msc_enrich_results_df$consq_name)
sv_msc_enrich_results_df$consq_name_f <- factor(sv_msc_enrich_results_df$consq_name_f, levels=sv_consequence_order)

rare_sv_msc_type_enrich_bonf_pass_pdf_path = file.path(fig2_out_dir, "rare_sv_consq_by_gene_z10_200kb_gene_bonf_pass_any.pdf")
pdf(rare_sv_msc_type_enrich_bonf_pass_pdf_path, width = 5,height = 4, useDingbats=FALSE)

sv_msc_enrich_fig <- ggplot(sv_msc_enrich_results_df, aes(y=consq_name_f, x=risk_ratio, color=vrnt_type)) +
  geom_vline(xintercept=1.0, linetype="solid", size=ticks_size) + 
  geom_point(size=3) + 
  ggtitle("") +
  geom_errorbar(aes(xmin=risk_ratio_lower, xmax=risk_ratio_upper),cex=1, width=0) + 
  scale_x_continuous(name = "Enrichment", expand=c(0.1, 0.1)) +
  scale_y_discrete(name = "") +
  facet_wrap(~vrnt_type, ncol = 1, as.table=TRUE, scales = "free", labeller=as_labeller(sv_class_labeller)) +
  scale_colour_manual(values = sv_vrnt_colors, guide="none") + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio > 1, "*", ""), x=risk_ratio_upper), hjust = -1, vjust = 0.5, size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio < 1, "*", ""), x=risk_ratio_lower), hjust = 1, vjust = 0.5, size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL))) 

theme.sv_class_enrich(sv_msc_enrich_fig)
dev.off()
theme.sv_class_enrich(sv_msc_enrich_fig)
rare_sv_msc_enrich_path_png <- file.path(fig2_out_dir, "rare_sv_consq_by_gene_z10_200kb_gene_bonf_pass_any.png")
ggsave(rare_sv_msc_enrich_path_png, width=5, height=4, dpi=360)
```

### Supplementary SV consequences that pass Bonferroni cutoff across z-score cutoffs 
```{r}
### Supplementary SV consequences across z-score cutoffs 
# all consequences that pass Bonferroni significance across z-scores

theme.sv_class_enrich <- function(p, base_size = 11, base_family = "Helvetica") {
  p <- p + theme_grey(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          panel.border = element_rect(color = "black", fill = NA, size = border_size),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_line(size=ticks_size),
          axis.ticks.length = unit(.15, "cm"),
          axis.title.x = element_text(size=base_size+2, family=base_family, face="plain", vjust=-1, margin = unit(c(2, 0, 0, 0), "mm"), color="black"),
          axis.title.y = element_text(size=base_size+2, family=base_family, face="plain", angle=90, margin = unit(c(0, 2, 0, 0), "mm"), color="black"),
          axis.text.y = element_text(size=base_size+2, family=base_family, face="plain", hjust=1, margin = unit(c(0, 1, 0, 0), "mm"), color="black"),
          axis.text.x = element_text(size=base_size+2, family=base_family, face="plain", hjust=0.5, margin = unit(c(1, 0, 0, 0), "mm"), color="black"),
          legend.key = element_blank(),
          legend.title = element_text(size=base_size+2, family = base_family, face="bold"),
          legend.text = element_text(size=base_size+2, family = base_family, face="plain"),
          legend.position = "none",
          complete = TRUE,
          strip.placement = "outside",
          strip.background = element_blank(),
          strip.text = element_text(size=base_size+2, family = base_family, face="plain", hjust=0, angle=0, vjust = 1),
          plot.title = element_text(hjust=0.5, size=base_size+2, family=base_family, face="bold"),
          panel.spacing = unit(0.2, "lines"))
  return(p)
}

sv_msc_enrich_results_path <-file.path(wkdir,'4_vrnt_enrich/enrich_results_mul_test/grouped_enrich/sv_consq_pass_bonf.tsv')
sv_msc_enrich_results_df <- read.csv(sv_msc_enrich_results_path, sep="\t")

sv_msc_enrich_results_df <- subset(sv_msc_enrich_results_df, sv_msc_enrich_results_df$z_cutoff %in% z_cutoffs_to_show)
sv_msc_enrich_results_df$z_cutoff_name <- as.character(sv_msc_enrich_results_df$z_cutoff_name)
sv_msc_enrich_results_df$z_cutoff_name <- factor(sv_msc_enrich_results_df$z_cutoff_name, levels=z_cutoffs_names)

y_axis_labels <- rev(unique(c('Upstream', 'Non-coding transcript','Coding','Transcript amplification','No predicted effect')))

sv_class_window_labeller <- c("DEL"="Deletions", 
                              "DUP"="Duplications", 
                              "INV"="Inversions", 
                              'Upstream'='Upstream', 
                              'Non-coding transcript'='Non-coding transcript',
                              'Coding'='Coding',
                              'Transcript amplification'='Transcript amplification',
                              'No predicted effect'='No predicted effect'
                              )

rare_sv_msc_type_enrich_bonf_pass_all_zcutoff_pdf_path = file.path(supp_fig8_out_dir, "rare_sv_consq_by_gene_z10_200kb_gene_bonf_pass_any_all_zcutoff.pdf")
pdf(rare_sv_msc_type_enrich_bonf_pass_all_zcutoff_pdf_path, width = 8,height = 7, useDingbats=FALSE)


sv_msc_all_enrich <- ggplot(sv_msc_enrich_results_df, aes(y=risk_ratio_pass, x=z_cutoff_name, color=vrnt_type)) +
  geom_hline(yintercept=1.0, linetype="solid", size=ticks_size) + 
  geom_point(size=3) + 
  ggtitle("") +
  geom_errorbar(aes(ymin=risk_lower_pass, ymax=risk_upper_pass),cex=1, width=0) + 
  scale_y_continuous(name = "Enrichment", expand=c(0.3, 0.3)) +
  scale_x_discrete(name = "Misexpression z-score theshold", labels=c("2", "10", "20", "30", "40")) +
  facet_wrap(vrnt_type~consq_name, ncol = 2,  scales = "free_y", as.table=TRUE, labeller=as_labeller(sv_class_window_labeller)) + 
  scale_colour_manual(name = "", 
                      values = sv_vrnt_colors, 
                      guide="none") + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True"  & risk_ratio_pass > 1, "*", ""), y=risk_upper_pass), hjust = 0.5, vjust = -0.25, size=4.5, fontface="bold", show.legend  = FALSE) + 
  geom_text(aes(label = ifelse(bonferroni_pass  == "True" & risk_ratio_pass < 1, "*", ""), y=risk_lower_pass), hjust = 0.5, vjust = 1.75, size=4.5, fontface="bold", show.legend  = FALSE) + 
  guides(fill = guide_legend(override.aes = list(label = NULL))) 

theme.sv_class_enrich(sv_msc_all_enrich)
dev.off()
theme.sv_class_enrich(sv_msc_all_enrich)
rare_sv_msc_type_enrich_bonf_pass_all_zcutoff_png_path = file.path(supp_fig8_out_dir, "rare_sv_consq_by_gene_z10_200kb_gene_bonf_pass_any_all_zcutoff.png")
ggsave(rare_sv_msc_type_enrich_bonf_pass_all_zcutoff_png_path, width=8, height=7, dpi=360)
```
